packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -58,24 +58,48 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
     protected readonly onWillSaveModelEmitter = new Emitter<WillSaveMonacoModelEvent>();
     readonly onWillSaveModel = this.onWillSaveModelEmitter.event;
 
+    private preferredEncoding: string | undefined = undefined;
+    private readonly defaultEncoding: string | undefined;
+
     constructor(
         protected readonly resource: Resource,
         protected readonly m2p: MonacoToProtocolConverter,
-        protected readonly p2m: ProtocolToMonacoConverter
+        protected readonly p2m: ProtocolToMonacoConverter,
+        options?: { encoding?: string | undefined }
     ) {
         this.toDispose.push(resource);
         this.toDispose.push(this.toDisposeOnAutoSave);
         this.toDispose.push(this.onDidChangeContentEmitter);
         this.toDispose.push(this.onDidSaveModelEmitter);
         this.toDispose.push(this.onWillSaveModelEmitter);
         this.toDispose.push(this.onDirtyChangedEmitter);
-        this.resolveModel = resource.readContents().then(content => this.initialize(content));
+        this.resolveModel = resource.readContents(options).then(content => this.initialize(content));
+        this.defaultEncoding = options && options.encoding ? options.encoding : undefined;
     }
 
     dispose(): void {
         this.toDispose.dispose();
     }
 
+    reopenWithEncoding(encoding: string) {
+        if (encoding === this.preferredEncoding || (!this.preferredEncoding && encoding === this.defaultEncoding)) {
+            return;
+        }
+        if (this.dirty) {
+            return;
+        }
+        this.preferredEncoding = encoding;
+        this.sync();
+    }
+
+    saveWithEncoding(encoding: string) {
+        // FIXME: need the ability from 'filesystem' to overwrite the encoding of file.
```

Comment:
could you explain what's missing?

Reply:
@svenefftinge Actually I figured out how to overwrite the encoding. Now, "Save with Encoding" is available as well. Here are two additional demos:

1. save with encoding (unchanged file)
![encoding-support-demo-3](https://user-images.githubusercontent.com/17487291/59331151-c2062600-8d25-11e9-9bc2-01f184edc6bc.gif)

2. save with encoding (dirty file)
![encoding-support-demo-4](https://user-images.githubusercontent.com/17487291/59331163-c92d3400-8d25-11e9-90f2-97fec05ab45d.gif)

What do you think?